snorkel/parser.py

```diff
@@ -10,12 +10,133 @@
 import sys
 import warnings
 from bs4 import BeautifulSoup
-from collections import namedtuple, defaultdict
+from collections import defaultdict
 from subprocess import Popen
 import lxml.etree as et
-from itertools import chain
+from snorkel import SnorkelBase, SnorkelSession, snorkel_postgres
+from sqlalchemy import Column, String, Integer, Text, ForeignKey
+from sqlalchemy.dialects import postgresql
+from sqlalchemy.orm import relationship, backref
+from sqlalchemy.types import PickleType
+
+
+class Corpus(SnorkelBase):
+    """
+    A Corpus holds a set of Documents and associated Sentences.
+    Default iterator is over (Document, Sentences) tuples
+    """
+
+    __tablename__ = 'corpus'
+    id = Column(Integer, primary_key=True)
+    name = Column(String)
+
+    def __repr__(self):
+        return "Corpus" + str((self.id, self.name))
+
+    def __iter__(self):
+        """Default iterator is over (document, sentence) tuples"""
+        for doc in self.documents:
+            for sentence in doc.sentences:
+                yield (doc, sentence)
+
+    def iter_docs(self):
+        for doc in self.documents:
+            yield doc
+
+    def iter_sentences(self):
+        for doc in self.documents:
+            for sentence in doc.sentences:
+                yield sentence
+
+    def get_docs(self):
+        return self.documents
+
+    def get_sentences(self):
+        return [sentence for doc in self.documents for sentence in doc.sentences]
+
+    def get_doc(self, id):
+        session = SnorkelSession.object_session(self)
+        return session.query(Document).filter(Document.id == id).one()
+
+    def get_sentence(self, id):
+        session = SnorkelSession.object_session(self)
+        return session.query(Document).filter(Sentence.id == id).one()
+
+    def get_sentences_in(self, doc_id):
+        session = SnorkelSession.object_session(self)
+        return session.query(Document).filter(Document.id == id).one().sentences
+
+
+class Document(SnorkelBase):
+    __tablename__ = 'document'
+    id = Column(Integer, primary_key=True)
+    corpus_id = Column(Integer, ForeignKey('corpus.id'))
+    corpus = relationship(Corpus, backref=backref('documents', uselist=True, cascade='delete,all'))
+    name = Column(String)
+    file = Column(String)
+    attribs = Column(PickleType)
```

Comment:
This `attribs` dict may contain an `xml.etree` object, which usually breaks pickle... we should check for issues here (specifically, see if this comes up when adding to `tutorial/CDR_tutorial`)